[MRG] Simulate spikes / action potentials

ryanhammonds opened the pull request:
This adds action potential cycles as the sum of two inverted asymmetrical gaussians:

```python
import matplotlib.pyplot as plt
import numpy as np
from neurodsp.sim.cycles import sim_ap_cycle

n_seconds = 1
fs = 1000

cycle_params = {'centers': (.25, .5), 'stds':(.15, .12),
                'alphas':(4, .2), 'heights': (15, -5)}

cyc = sim_ap_cycle(n_seconds, fs, **cycle_params)

plt.plot(cyc)
```
![ap_cycle](https://user-images.githubusercontent.com/34786005/115317653-e0fb4d80-a130-11eb-8a85-7bf410546dfa.png)

ryanhammonds commented:
This func can be used with scipys curve_fit to fit  segmented APs from real data (bycycle-tools/bycycle#100). The resulting 6 params could be useful (i.e. dimensionality reduction for APs) and included in df_features returned from bycycle.Spikes.fit().
![ap_cycle_fit](https://user-images.githubusercontent.com/34786005/115329862-caacbc00-a147-11eb-802c-9ad0874e2ed7.png)

TomDonoghue commented on neurodsp/sim/cycles.py:
```diff
@@ -243,7 +253,89 @@ def sim_gaussian_cycle(n_seconds, fs, std):
+    return cycle
+
+
+def sim_skewed_gaussian_cycle(n_seconds, fs, center, std, alpha, height=1):
```
The parameter `alpha` isn't described in the docstring

TomDonoghue commented on neurodsp/sim/cycles.py:
```diff
@@ -230,6 +236,10 @@ def sim_gaussian_cycle(n_seconds, fs, std):
     std : float
         Standard deviation of the gaussian kernel, in seconds.
+    center : float, optional, default: 0.5
+        The center of the gaussian.
```
This definition isn't super clear (it's not obvious if it's about x or y axis), and I think the description in `sim_skewed_gaussian` is clearer and could be copied here.

TomDonoghue commented:
This looks super cool! 

I had a quick scan of the code, left a coupe comments, and here's a couple quick "higher-level" questions:
- It's not clear to me that the AP sim is a _cycle_, per se. Would it fit better in `sim.transients`?
- I presume changing the gaussian code to not use scipy is about getting more control over height, center, etc? Setting `height` feels like it overlaps with normalizing the signal, and `center` feels a little weird to have in the cycles, because it makes looks like it 'cuts off' the edge of a gaussian? Am I right in thinking that one would probably never want to set `center` to a custom value because it would create weird concatenation artifacts? 

Between the two comments, I think it's possible it might be a cleaner split to put some stuff in transients, that might work well for generating APs, but might not be quite the same as what one wants for a cycle? What do you think?

ryanhammonds commented:
Thanks for the review!

- I think it makes sense to refactor into `transients` with an alias in `cycles` so it can still be used with sim_bursty_oscillation.
- The main reason for the sim_gaussian_cycle changes was to have control over the cycle center, which is important only for skewed cycles. Here's an example of how center makes a difference with skewed cycles:

![skew_center](https://user-images.githubusercontent.com/34786005/115774615-a8dd5000-a366-11eb-871c-c57537589361.png)

The `height` param is a bit different from `variance`. I'm not sure there is a clear mapping between these params.
![variance_v_height](https://user-images.githubusercontent.com/34786005/115774516-8b0feb00-a366-11eb-83fb-5151b4d17929.png)

ryanhammonds commented:
I've update the code here based on the review. I also update the ap func to take iterables for simulation parameters. This is because 3 gaussians may desired in some cases (i.e. the sum of conductive, sodium and potassium currents).

```python
from neurodsp.sim.cycles import sim_ap_cycle
import matplotlib.pyplot as plt

cycle = sim_ap_cycle(1, 100, (.5, .3, .6), (.1, .1, .1), (0, -1, 1), (-5, 1, 2.5))

plt.plot(cycle)
```
![example_double_skew](https://user-images.githubusercontent.com/34786005/116180722-3fdf3a80-a6ce-11eb-8dd5-de2702a9dd60.png)

ryanhammonds commented:
@TomDonoghue I made a final round of updates. There have been several changes since the last time you reviewed, but it should be ready to merge. Let me know if you'd like to look over the changes, otherwise I'll merge.
